openupgrade_scripts/scripts/account/17.0.1.2/post-migration.py

```diff
@@ -0,0 +1,535 @@
+            )
+
+
+def _account_payment_term_migration(env):
+    """
+    Switch balance lines to percent and compute the remaining percentage, and convert
+    old multiple column system to the new delay_type + nb_days.
+
+    https://github.com/odoo/odoo/pull/110274
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_payment_term_line
+        SET value = 'percent',
+            value_amount = 100.0 - COALESCE(percentages.percentage, 0)
+        FROM (
+            SELECT
+                payment_id,
+                SUM(
+                    CASE WHEN l.value='percent' THEN l.value_amount
+                    ELSE 0 END
+                ) percentage
+            FROM account_payment_term_line l
+            GROUP BY payment_id
+        ) percentages
+        WHERE
+        value = 'balance' AND
+        percentages.payment_id = account_payment_term_line.payment_id
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_payment_term_line
+        SET delay_type = CASE
+            WHEN end_month AND COALESCE(months, 0) = 0
+                AND COALESCE(days, 0) = 0
+                THEN 'days_after_end_of_month'
+            WHEN end_month AND months = 1 AND COALESCE(days, 0) = 0
+                THEN 'days_after_end_of_next_month'
+            WHEN end_month AND COALESCE(months, 0) <= 1 AND days > 0
+                THEN 'days_end_of_month_on_the'
+            ELSE 'days_after'
+        END,
+        nb_days = CASE
+            WHEN end_month AND months <= 1
+                THEN COALESCE(days, 0) + COALESCE(days_after, 0)
+            ELSE
+                COALESCE(months, 0)*30 + COALESCE(days, 0) +
+                COALESCE(days_after, 0)
+        END
+        """,
+    )
+
+
+def _account_payment_term_early_payment_discount(env):
+    """Only payment terms with one line and the early payment discount activated are
+    valid now, so we are going to discard other previous configurations.
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+        WITH sub AS (
+            SELECT * FROM (
+                SELECT *,
+                    row_number() over (partition BY payment_id ORDER BY id) AS rnum
+                FROM account_payment_term_line
+            ) t
+            WHERE t.rnum = 1
+            AND t.discount_days IS NOT NULL
+            AND t.discount_percentage > 0
+        )
+        UPDATE account_payment_term apt
+        SET early_discount = True,
+            discount_days = sub.discount_days,
+            discount_percentage = sub.discount_percentage
+        FROM sub
+        WHERE sub.payment_id = apt.id
+        """,
+    )
+
+
+def convert_from_company_dependent(
+    env,
+    model_name,
+    origin_field_name,
+    destination_field_name,
+    origin_id_column_name,
+    model_table_name=None,
+):
+    """
+    Move a company-dependent field back to the model table.
```

Comment:
I think it could be useful to move this function to `openupgradelib` to improve reusability. It could serve as a base utility for scenarios requiring the conversion of property fields to JSON during migrations to version 18.

Reply:
that's a very different logic, you'll find the v18 function [here](https://github.com/OCA/openupgradelib/pull/396)

By now I've seen most of the migrations for v17 and didn't see another module where the case we have in the account module happens.